### PR TITLE
[SC] Merge dev-sc

### PR DESCRIPTION
**Pull request by rowandh**

### CODE REVIEW

File: src/Stratis.Bitcoin.Features.SmartContracts/Rules/TxOutSmartContractExecRule.cs [hunk 1]
```diff
@@ -32,7 +33,11 @@ public void CheckTransaction(MempoolValidationContext context)
 
         private void CheckTransaction(Transaction transaction)
         {
-            var smartContractExecCount = transaction.Outputs.Count(o => SmartContractScript.IsSmartContractExec(o.ScriptPubKey));
+            int smartContractExecCount = transaction.Outputs.Count(o => o.ScriptPubKey.IsSmartContractExec());
+
+            if ((transaction.IsCoinBase)  && smartContractExecCount >= 1)
+                new ConsensusError("smartcontractexec-in-coinbase", "coinbase or coinstake contains smartcontractexec output").Throw();
```
Comment (codingupastorm) [inline src/Stratis.Bitcoin.Features.SmartContracts/Rules/TxOutSmartContractExecRule.cs]:
```suggestion
                new ConsensusError("smartcontractexec-in-coinbase", "coinbase contains smartcontractexec output").Throw();
```

File: src/Stratis.Bitcoin.Features.SmartContracts/Rules/TxOutSmartContractExecRule.cs [hunk 2]
```diff
@@ -32,7 +33,11 @@ public void CheckTransaction(MempoolValidationContext context)
 
         private void CheckTransaction(Transaction transaction)
         {
-            var smartContractExecCount = transaction.Outputs.Count(o => SmartContractScript.IsSmartContractExec(o.ScriptPubKey));
+            int smartContractExecCount = transaction.Outputs.Count(o => o.ScriptPubKey.IsSmartContractExec());
+
+            if ((transaction.IsCoinBase)  && smartContractExecCount >= 1)
```
Comment (codingupastorm) [inline src/Stratis.Bitcoin.Features.SmartContracts/Rules/TxOutSmartContractExecRule.cs]:
```suggestion
            if ((transaction.IsCoinBase)  && smartContractExecCount > 0)
```

File: src/Stratis.SmartContracts.CLR/Compilation/ContractDecompiler.cs [hunk 3]
```diff
@@ -17,10 +17,10 @@ public static Result<IContractModuleDefinition> GetModuleDefinition(byte[] bytes
...
             try
             {
-                var moduleDefinition = ModuleDefinition.ReadModule(stream, new ReaderParameters { AssemblyResolver = resolver });
+                var moduleDefinition = ModuleDefinition.ReadModule(stream, new ReaderParameters { AssemblyResolver = resolver});
                 return Result.Ok<IContractModuleDefinition>(new ContractModuleDefinition(moduleDefinition, stream));
             }
-            catch(BadImageFormatException e)
+            catch (Exception e) // Ideally we would only catch BadImageFormatException but unfortunately Cecil can be unreliable.
             {
                 return Result.Fail<IContractModuleDefinition>("Invalid bytecode:" + e.Message);
```
Comment (codingupastorm) [inline src/Stratis.SmartContracts.CLR/Compilation/ContractDecompiler.cs]: Check if this gets logged

File: src/Stratis.SmartContracts.CLR/ContractModuleDefinition.cs [hunk 4]
```diff
@@ -48,26 +49,51 @@ public TypeDefinition ContractType
...
         {
-            this.ModuleDefinition = rewriter.Rewrite(this.ModuleDefinition);
+            try
+            {
+                this.ModuleDefinition = rewriter.Rewrite(this.ModuleDefinition);
+                return true;
+            }
+            catch (Exception)
+            {
+                return false;
```
Comment (codingupastorm) [inline src/Stratis.SmartContracts.CLR/ContractModuleDefinition.cs]: Should log here, and on line 79 below in this file

File: src/Stratis.SmartContracts.CLR/ReflectionVirtualMachine.cs [hunk 5]
```diff
@@ -205,7 +228,16 @@ private static VmExecutionResult GetInvocationVmErrorResult(IContractInvocationR
...
-            IContract contract = Contract.CreateUninitialized(type, contractState, address);
+            IContract contract;
+
+            try
+            {
+                contract = Contract.CreateUninitialized(type, contractState, address);
+            }
+            catch (Exception)
+            {
+                return Result.Fail<IContract>("Error instantiating contract instance");
```
Comment (codingupastorm) [inline src/Stratis.SmartContracts.CLR/ReflectionVirtualMachine.cs]: Log the error or better, include exception in result message

File: src/Stratis.SmartContracts.CLR/ContractModuleDefinition.cs (hunk 4 above)
Comment (rowandh) [reply src/Stratis.SmartContracts.CLR/ContractModuleDefinition.cs]: I've moved all the exception handling to `ReflectionVirtualMachine` and added logging.

**Pull request merged**